selfservice/hook/web_hook.go

```diff
@@ -0,0 +1,187 @@
+	"github.com/ory/kratos/selfservice/flow/verification"
+	"github.com/ory/kratos/session"
+	"github.com/ory/kratos/x"
+	"io"
+	"net/http"
+)
+
+var _ registration.PostHookPostPersistExecutor = new(WebHook)
+var _ verification.PostHookExecutor = new(WebHook)
+
+type (
+	webHookDependencies interface {
+		x.LoggingProvider
+	}
+
+	emptyAuthConfig struct {}
+
+	basicAuthConfig struct {
+		User     string
+		Password string
+	}
+
+	apiKeyConfig struct {
+		Name  string
+		Value string
+		In    string
+	}
+
+	AuthConfig interface {
+		apply(req *http.Request)
+	}
+
+	Auth struct {
+		Type       string
+		AuthConfig AuthConfig      `json:"-"`
+		RawConfig  json.RawMessage `json:"config"`
+	}
+
+	webHookConfig struct {
+		Method string
+		Url    string
+		Body   string
+		Auth   Auth
+	}
+
+	WebHook struct {
+		r webHookDependencies
+		c json.RawMessage
+	}
+)
+
+func (c *emptyAuthConfig) apply(_ *http.Request) {}
+
+func (c *webHookConfig) Unmarshal(bytes []byte) error {
+	if err := json.Unmarshal(bytes, c); err != nil {
+		return err
+	}
+
+	if len(c.Auth.Type) == 0 {
+		c.Auth.AuthConfig = &emptyAuthConfig{}
+	}
+	return nil
+}
+
+func (c *basicAuthConfig) apply(req *http.Request) {
+	req.SetBasicAuth(c.User, c.Password)
+}
+
+func (c *apiKeyConfig) apply(req *http.Request) {
+	switch c.In {
+	case "cookie":
+		req.AddCookie(&http.Cookie{ Name: c.Name, Value: c.Value })
+	default:
+		req.Header.Set(c.Name, c.Value)
+	}
+}
+
+func (a *Auth) UnmarshalJSON(bytes []byte) error {
+	type auth Auth
+	err := json.Unmarshal(bytes, (*auth)(a))
+	if err != nil {
+		return err
+	}
+	switch a.Type {
+	case "basic-auth":
+		var authConfig basicAuthConfig
+		err = json.Unmarshal(a.RawConfig, &authConfig)
+		if err != nil {
+			return err
+		}
+		a.AuthConfig = &authConfig
+	case "api-key":
+		var authConfig apiKeyConfig
+		err = json.Unmarshal(a.RawConfig, &authConfig)
+		if err != nil {
+			return err
+		}
+		a.AuthConfig = &authConfig
+	default:
+		return fmt.Errorf("unknown auth type %v", a.Type)
+	}
+	return nil
+}
+
+func NewWebHook(r webHookDependencies, c json.RawMessage) *WebHook {
+	return &WebHook{r: r, c: c}
+}
+
+func (e *WebHook) ExecutePostVerificationHook(_ http.ResponseWriter, _ *http.Request, f *verification.Flow, s *session.Session) error {
+	return e.executeWebHook(f, s)
+}
+
+func (e *WebHook) ExecutePostRegistrationPostPersistHook(_ http.ResponseWriter, _ *http.Request, f *registration.Flow, s *session.Session) error {
+	return e.executeWebHook(f, s)
+}
+
+func (e *WebHook) executeWebHook(f interface{}, s interface{}) (err error) {
+	conf := &webHookConfig{}
+	if err := conf.Unmarshal(e.c); err != nil {
+		return err
+	}
+	var body io.Reader
+	if len(conf.Body) != 0 {
+		body, err = e.createBody(conf.Body, f, s)
+		if err != nil {
+			return fmt.Errorf("failed to create web hook body: %w", err)
+		}
+	}
+	if err = e.doHttpCall(conf, body); err != nil {
+		return fmt.Errorf("failed to call web hook %w", err)
```

Comment:
One could add an additional strategy here, defining how the actual execution of the Web hook happens, like synchronously, or asynchronously, with or without retries, whether the response should somehow be handled or not, etc.